File: rest-service/api/src/main/java/org/veupathdb/service/multiblast/model/internal/JobStatus.java
package org.veupathdb.service.multiblast.model.internal;

import java.util.Optional;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

public enum JobStatus
{
  Queued("queued"),
  InProgress("in-progress"),
  Errored("errored"),
  Completed("completed"),
  Expired("expired");

  private final String value;

  JobStatus(String name) {
    this.value = name;
  }

  @JsonValue
  public String getValue() {
    return value;
  }

  public static Optional<JobStatus> fromString(String name) {
    for (var e : values())
      if (e.value.equals(name))
        return Optional.of(e);

    return Optional.empty();
  }

  @JsonCreator
  public static JobStatus unsafeFromString(String name) {
    return fromString(name)
      .orElseThrow(() -> new IllegalArgumentException("Unrecognized JobStatus value: " + name));
  }
}

File: lib-jvm-mblast-sdk/src/main/kotlin/org/veupathdb/lib/mblast/sdk/common/model/JobStatus.kt
package org.veupathdb.lib.mblast.sdk.common.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.TextNode
import java.lang.IllegalArgumentException

enum class JobStatus {
  Queued,
  InProgress,
  Complete,
  Failed,
  Expired,
  ;

  @JsonValue
  fun toJson() =
    when (this) {
      Queued     -> TextNode("queued")
      InProgress -> TextNode("in-progress")
      Complete   -> TextNode("complete")
      Failed     -> TextNode("failed")
      Expired    -> TextNode("expired")
    }

  companion object {

    @JvmStatic
    @JsonCreator
    fun fromJson(js: JsonNode): JobStatus {
      if (!js.isTextual)
        throw IllegalArgumentException()

      return when (js.textValue()) {
        "queued"      -> Queued
        "in-progress" -> InProgress
        "complete"    -> Complete
        "failed"      -> Failed
        "expired"     -> Expired
        else          -> throw IllegalArgumentException()
      }
    }
  }
}